### ISSUE TITLE
Connection issue 

### ISSUE THREAD

**Issue by njsurf73** [FYSETC/FYSETC-PortableInputShaper#5]
When I copy the firmware to the rp2040 it will not be recognized when I unplug and move to my PC running klipper. In order to make it work, I have to connect to the PC running klipper and then copy the firmware to the rp2040. And every time I restart I have to copy the firmware to the 2040 again on the PC I am using to run klipper.

**Comment by DaVinci-10** [comment]
What is the os on the PC running klipper?
I am currently using Debian desktop version to run klipper. And when logged in. I first stop klipper. Unplug usb, hold the boot button while plugging in usb. It pops a box on the laptop asking if it should open the drive in Files app. Then I just navigate to klipper/out folder copy klipper.uf2 and paste onto the rp2040 drive...( copy paste, dont drag and drop)
 instantly the whole drive disappears...
Restart klipper and done. I've done it twice now no issues.
Can you post a screenshot of the make screen?

**Comment by njsurf73** [comment]
My issue is, I have 3 printers running on one laptop. If I restart the laptop, I have to copy the firmware to the 2040 allover again to have it recognized. Otherwise it doesn't show back up as a mcu

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by DaVinci-10** [comment]
What shows up in a lsusb, or ls /dev/serial/by-id/  (when it works and when it doesnt)
also can you also post your pis.cfg, maybe even a klippy.log
what frontend mainsail or fluidd? 
which printer is configured to use it? 7125 or higher or all and you just uncomment the one you have it on... ?

**Comment by njsurf73** [comment]
I uncomment the printers not in use. I will post a log when I get home.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by fryc88** [comment]
![image](https://user-images.githubusercontent.com/50572708/233833502-8a929a3d-48db-4e43-b674-3b1850bd3923.png)

That setting, bootloader offset apparently solved my issue like yours, with disappearing device after connecting it from pc to my rpi with klipper. When this manual has been made, apparently klipper haven't had this option in make menuconfig.

**Comment by njsurf73** [comment]
When I do this I get a .bin not a .uf2 file in klipper/outDid you change that or just load that On Apr 23, 2023 6:10 AM, fryc88 ***@***.***> wrote:

That setting, bootloader offset apparently solved my issue like yours, with disappearing device after connecting it from pc to my rpi with klipper. When this manual has been made, apparently klipper haven't had this option in make menuconfig.

**Comment by DaVinci-10** [comment]
the current klipper makemenu creates the .uf2 after compiling the .bin for the 2040
![image](https://user-images.githubusercontent.com/1100376/233872780-218c1f57-6c0b-4dc1-99e4-96453a63ebe2.png)

Version: v0.11.0-180-g011b4e39
  Building rp2040 stage2 out/stage2.o
  Preprocessing out/src/rp2040/rp2040_link.ld
  Linking out/klipper.elf
  Building out/lib/rp2040/elf2uf2/elf2uf2
  Creating uf2 file out/klipper.uf2
[✓ OK] Firmware built!

**Comment by njsurf73** [comment]
> 

No matter how I build it, when I select the bootloader offset it only compiles a .bin
Using the makemenu and kiauh 
It will only build it to .bin

**Comment by DaVinci-10** [comment]
bootloader offset should be set to none...
![image](https://user-images.githubusercontent.com/1100376/234173439-e755fdf2-ed47-40b3-857a-fe8a54d0d3c5.png)
should look like this

```G3rKo — Today at 6:14 PM
CD klipper, make menuconfig,  select RP2040 ,no bootloader, connection usb,
Esc, yes, 
make -j4
should compile out/klipper.uf2
If select bootloader then no uf2 file will be generated.!
and you may need write access  "sudo cp out/klipper.uf2 /media..../"
And in PIS.config u need to replace the serial=/dev/serial/by-id/   with your id
get the id by "ls /dev/serial/by-id"```
from a discord discussion... 

can you capture this kiauh screen

```Version: v0.11.0-180-g011b4e39
  Building rp2040 stage2 out/stage2.o
  Preprocessing out/src/rp2040/rp2040_link.ld
  Linking out/klipper.elf
  Building out/lib/rp2040/elf2uf2/elf2uf2
  Creating uf2 file out/klipper.uf2
[✓ OK] Firmware built!
/=======================================================\
|     ~~~~~~~~~~~~~ [ Advanced Menu ] ~~~~~~~~~~~~~     |
|-------------------------------------------------------|
| Klipper & API:          | Mainsail:                   |
|  1) [Rollback]          |  6) [Theme installer]       |
|                         |                             |
| Firmware:               | System:                     |
|  2) [Build only]        |  7) [Change hostname]       |
|  3) [Flash only]        |                             |
|  4) [Build + Flash]     | Extras:                     |
|  5) [Get MCU ID]        |  8) [G-Code Shell Command]  |
|-------------------------------------------------------|
|                       B) « Back                       |
\=======================================================/
####### Perform action:```

**Comment by njsurf73** [comment]
> bootloader offset should be set to none... ![image](https://user-images.githubusercontent.com/1100376/234173439-e755fdf2-ed47-40b3-857a-fe8a54d0d3c5.png) should look like this
> 
> ```
> CD klipper, make menuconfig,  select RP2040 ,no bootloader, connection usb,
> Esc, yes, 
> make -j4
> should compile out/klipper.uf2
> If select bootloader then no uf2 file will be generated.!
> and you may need write access  "sudo cp out/klipper.uf2 /media..../"
> And in PIS.config u need to replace the serial=/dev/serial/by-id/   with your id
> get the id by "ls /dev/serial/by-id"```
> from a discord discussion... 
> 
> can you capture this kiauh screen
> 
> ```Version: v0.11.0-180-g011b4e39
>   Building rp2040 stage2 out/stage2.o
>   Preprocessing out/src/rp2040/rp2040_link.ld
>   Linking out/klipper.elf
>   Building out/lib/rp2040/elf2uf2/elf2uf2
>   Creating uf2 file out/klipper.uf2
> [✓ OK] Firmware built!
> /=======================================================\
> |     ~~~~~~~~~~~~~ [ Advanced Menu ] ~~~~~~~~~~~~~     |
> |-------------------------------------------------------|
> | Klipper & API:          | Mainsail:                   |
> |  1) [Rollback]          |  6) [Theme installer]       |
> |                         |                             |
> | Firmware:               | System:                     |
> |  2) [Build only]        |  7) [Change hostname]       |
> |  3) [Flash only]        |                             |
> |  4) [Build + Flash]     | Extras:                     |
> |  5) [Get MCU ID]        |  8) [G-Code Shell Command]  |
> |-------------------------------------------------------|
> |                       B) « Back                       |
> \=======================================================/
> ####### Perform action:```
> ```

Yes. I can do that fine and get the correct firmware, but once I flash it to the pis, unless it is connected to my host it won't be recognized. If I flash it on the host, as soon as I restart it disappears. I was trying the boot offset because it was recommended as a fix.

**Comment by DaVinci-10** [comment]
(sorry if this sounds over simplified, but the flashing procedure is simple file copy and paste, you dont flash it per se. ie you dont use the standard flash procedure like you do the printer mcu. 
It's easiest to do on a desktop flavor of linux, because as soon as you plug the usb in while holding the boot button down, A Jump (flash) drive pops up on the display, simply open and copy and paste the file from home/klipper/out. And as soon as you copy the file, the 2040 automaticly flashes itself internally and reboots. thats why the flash drive disappears.

**Comment by njsurf73** [comment]
> (sorry if this sounds over simplified, but the flashing procedure is simple file copy and paste, you dont flash it per se. ie you dont use the standard flash procedure like you do the printer mcu. It's easiest to do on a desktop flavor of linux, because as soon as you plug the usb in while holding the boot button down, A Jump (flash) drive pops up on the display, simply open and copy and paste the file from home/klipper/out. And as soon as you copy the file, the 2040 automaticly flashes itself internally and reboots. thats why the flash drive disappears.

yes, i have used a pico and many varients of an rp2040. i understand how to flash it. The problem is once flashed it will be ok as long as its flashed from the host and remains powered on and plugged in. once it loses power or is unplugged, it reverts to not connecting as an mcu and has to be flashed again. there is a fault with the PIS not the process.

**Comment by DaVinci-10** [comment]
Ok, I get that... so my next quesiton is:
After a hard boot, does it appear as a drive on the host like you held the boot button?
and more for the developer I guess, but does a soft reset after a flash ignore the boot button... ie would stuck boot button exibit that behavior

**Comment by njsurf73** [comment]
No. It shows up on lsusb but as a connected device. It doesn't show up by /serial. Have to hold the boot button down again and reconnect to get it flashable.

**Comment by SrgntBallistic** [comment]
I'm not having issues with the PIS not connecting
I just move the firmware over to the pico from the Pi

Disconnect the pico
Reconnect the pico while holding the boot button
run `> lsusb`
The Pico should show up as "RP2 Boot"
![image](https://github.com/FYSETC/FYSETC-PortableInputShaper/assets/25190883/48c50e08-47a2-41cb-92e4-66c3d2c63364)
 cd into Klipper
` > make menuconfig`
` > make`
` > sudo mount /dev/sda1 /mnt`
` > sudo cp out/klipper.uf2 /mnt`
` > sudo umount /mnt`
` > ls /dev/serial/by-id`
![image](https://github.com/FYSETC/FYSETC-PortableInputShaper/assets/25190883/f87d907a-d482-4b7d-8729-3496579deadf)

Finally disconnect and reconnect the Pico
That sticks for me and I can reconnect and rerun IS

This at least takes the process of connecting to a PC out of the flow

**Comment by njsurf73** [comment]
Unfortunately it doesn't stick for me. I have to reflash every time it disconnects. I put together my own, and it's working as expected

**Comment by WaschBaer0395** [comment]
how did you fix yours ?? im having the same issues, as soon as it disconnects after copying over the firmware file, it looses it and wont get recognized ... so i cant even use it since i move it over from my pc, disconnect it and then its gone again .. blank

**Comment by njsurf73** [comment]
> soon

I didn't, I binned it and got a new one. I was never able to get it the flash to be persistent.
